myia/pipeline/resources.py

```diff
@@ -149,25 +149,41 @@ def _on_drop_node(self, event, node):
 class InferenceResource(Partializable):
     """Performs inference and monomorphization."""
 
-    def __init__(self, resources, constructors, context_class):
+    def __init__(self, resources, constructors, max_stack_depth):
         """Initialize an InferenceResource."""
+        self.resources = resources
         self.manager = resources.manager
-        self.context_class = context_class
         self.constructors = constructors
+        self.max_stack_depth = max_stack_depth
         self.engine = InferenceEngine(
             resources,
             constructors=self.constructors,
-            context_class=self.context_class,
+            max_stack_depth=self.max_stack_depth,
+        )
+        self.live = LiveInferenceEngine(
+            resources,
+            constructors=self.constructors,
         )
 
-    def infer(self, graph, argspec, outspec=None, clear=False):
+    def infer_incremental(self):
+        """Perform inference."""
+        tracker = self.resources.tracker
+        if not tracker.activated:
+            return
+        mng = self.manager
+        todo = tracker.todo
+        while todo:
+            nodes = [node for node in todo
+                     if node.abstract is None and node in mng.all_nodes]
```

Comment:
@breuleux 
I am curious about this while loop. I noticed that for all existing tests, during second iteration of the loop, nodes in the todo list already have a type. So self.live.run() is executed with an empty list as an argument.

Is this while loop added with some particular use case in mind, or for some future features?

Reply:
You know, I think you're right, this loop is pointless.

Originally I think it was possible for new nodes to be added as a side-effect of the incremental inference, or I thought it might be possible, I'm not sure. In any case, I put it there at some point then never questioned its relevance again, so thanks for pointing it out!

Reply:
oh, I see. Thanks!